### PR TITLE
GrainTracker refactoring

### PR DESCRIPTION
**Pull request by permcody**
Memory cleanup and cosmetic loop refectoring

refs #6711

### CODE REVIEW

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 1]
```diff
@@ -266,12 +272,18 @@ void FeatureFloodCount::communicateAndMerge()
...
 
+  // Free up as much memory as possible here before we do global communication
+  clearDataStructures();
+
   /**
    * Each processor needs information from all other processors to create a complete
    * global feature map.
    */
+  _communicator.barrier();
+  Moose::perf_log.push("allgather_packed_range()", "FeatureFloodCount");
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Will we ever get integer handles for perf_log?

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 2]
```diff
@@ -498,17 +510,17 @@ FeatureFloodCount::deserialize(std::vector<std::string> & serialized_buffers)
   std::istringstream iss;
 
   mooseAssert(serialized_buffers.size() == _app.n_processors(), "Unexpected size of serialized_buffers: " << serialized_buffers.size());
-
-  for (auto rank = decltype(_n_procs)(0); rank < serialized_buffers.size(); ++rank)
+  auto rank = processor_id();
+  for (decltype(rank) proc_id = 0; proc_id < serialized_buffers.size(); ++proc_id)
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Wait, won't this break if `processor_id()` ever returned `const int`?

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 1 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Ask libMesh.

**Comment by dschwen** [comment]
Looks fine to me, :+1: for all the for loop updates.

**Pull request merged**